pkg/http/server_test.go

```diff
@@ -97,3 +97,66 @@ func TestNewServer(t *testing.T) {
 		}
 	}
 }
+
+var tlsCipherSuitesTests = []struct {
+	ciphers    []uint16
+	shouldFail bool
+}{
+	{ciphers: nil, shouldFail: false},
+	{ciphers: defaultCipherSuites, shouldFail: false},
+	{
+		ciphers: []uint16{ // unsupported cipher suites
+			tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,    // Go stack contains (some) countermeasures against timing attacks (Lucky13)
+			tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256, // No countermeasures against timing attacks
+			tls.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,    // Go stack contains (some) countermeasures against timing attacks (Lucky13)
+			tls.TLS_ECDHE_ECDSA_WITH_RC4_128_SHA,        // Broken cipher
+			tls.TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA,     // Sweet32
+			tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,      // Go stack contains (some) countermeasures against timing attacks (Lucky13)
+			tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256,   // No countermeasures against timing attacks
+			tls.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,      // Go stack contains (some) countermeasures against timing attacks (Lucky13)
+			tls.TLS_ECDHE_RSA_WITH_RC4_128_SHA,          // Broken cipher
+
+			// all RSA-PKCS1-v1.5 ciphers are disabled - danger of Bleichenbacher attack variants
+			tls.TLS_RSA_WITH_3DES_EDE_CBC_SHA,   // Sweet32
+			tls.TLS_RSA_WITH_AES_128_CBC_SHA,    // Go stack contains (some) countermeasures against timing attacks (Lucky13)
+			tls.TLS_RSA_WITH_AES_128_CBC_SHA256, // No countermeasures against timing attacks
+			tls.TLS_RSA_WITH_AES_256_CBC_SHA,    // Go stack contains (some) countermeasures against timing attacks (Lucky13)
+			tls.TLS_RSA_WITH_RC4_128_SHA,        // Broken cipher
+
+			tls.TLS_RSA_WITH_AES_128_GCM_SHA256, // Disabled because of RSA-PKCS1-v1.5 - AES-GCM is considered secure.
+			tls.TLS_RSA_WITH_AES_256_GCM_SHA384, // Disabled because of RSA-PKCS1-v1.5 - AES-GCM is considered secure.
+		},
+		shouldFail: true,
+	},
+}
+
+func TestTLSCiphers(t *testing.T) {
+	certificate, err := getTLSCert()
+	if err != nil {
+		t.Fatalf("Unable to parse private/certificate data. %v\n", err)
+	}
+	port := getNextPort()
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintf(w, "Hello, world")
+	})
+	server := NewServer([]string{"127.0.0.1:" + port}, handler, &certificate)
+	go func() { panic(server.Start()) }()
+	defer server.Shutdown()
+	for i, test := range tlsCipherSuitesTests {
+		client := http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{
+					InsecureSkipVerify: true,
+					CipherSuites:       test.ciphers,
```

Comment:
Actually we want to run http server with testCase.ciphers and making a client call to check whether the client succeeds or not.  The tests are in server point of view.

Reply:
That makes not much sense because we would check whether the standard go http client can connect to an "insecure" configured server instead of whether the server refuses such "insecure" requests.
Following your approach last test-case would succeed instead of fail - that's not what we want.

Reply:
unable to understand here.  As we support overriding ciphers by setting server.TLSConfig, What exact test we do here for this case?

Reply:
We test that 
1. The server accepts TLS connections when the client supports at least one of the "supported ciphers".
2. The server refuses TLS connections when the client supports only "unsupported ciphers".

>Actually we want to run http server with testCase.ciphers

No, we don't want to do that. The reason is that we would test that:
1. The server accepts TLS connections when the client supports only "supported ciphers".
2. The server accepts TLS connections when the client supports only "unsupported ciphers".

Reply:
@aead I don't know whether I am conveying correctly.  We need a test case exactly how current master behaves.

Reply:
The point is, that this PR changes the behavior of the server. So we have to test that:
1. The server accepts a TLS connection if the client supports at least one cipher of the "supported ciphers" (`defaultCipherSuites`)
2. The server rejects a TLS connection if the client does not support any cipher of the "supported ciphers" (`defaultCipherSuites`)
 
The TLSConfig: 
```
httpServer.TLSConfig = &tls.Config{
	PreferServerCipherSuites: true,
	MinVersion:               tls.VersionTLS12,
	NextProtos:               []string{"http/1.1", "h2"},
}
is not used by the minio server anymore (when this PR is merged). That's the point of this PR - Otherwise the server would support ciphers like 3DES_CBC_SHA - we want to remove them.
```

>To this server, there is no error http client gets for any cipher suite.

True, but exactly this behavior is changed by this PR. However the default Go http client can still connect. (That's the test-case: `{cipher: nil, shouldFail: false}`

Or do I miss your point here?

Reply:
> is not used by the minio server anymore (when this PR is merged). That's the point of this PR - Otherwise the server would support ciphers like 3DES_CBC_SHA - we want to remove them.

This `http` package is not limiting to `minio` only (that's also reason it made as separate package).  If someone uses `http` package for his project, it should be complete, so we have to test
```go
httpServer.TLSConfig = &tls.Config{
	PreferServerCipherSuites: true,
	MinVersion:               tls.VersionTLS12,
	NextProtos:               []string{"http/1.1", "h2"},
}
```

Reply:
I added a test for that. However I don't think it's very useful to test this behavior because the cipher negotiation is done by the `crypto/tls` package and it's not really useful to test the standard library.